grpc-client-server/prober_server/main_test.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func dialer() func(context.Context, string) (net.Conn, error) {
+	listener := bufconn.Listen(1024 * 1024)
+
+	server := grpc.NewServer()
+
+	pb.RegisterProberServer(server, &Server{})
+
+	go func() {
+		if err := server.Serve(listener); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	return func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+}
+
+func TestDoProbes(t *testing.T) {
+	tests := map[string]struct {
+		req       *pb.ProbeRequest
+		err       error
+		failedReq int32
+	}{
+		"simple": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "https://www.google.com",
+				NumberOfRequests: 1,
+			},
+			failedReq: 0,
+		},
+		"error": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "http://musterror:8080",
+				NumberOfRequests: 3,
+			},
+			failedReq: 3,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithTransportCredentials(insecure.NewCredentials()))
+			require.NoError(t, err)
+			defer conn.Close()
+
+			client := pb.NewProberClient(conn)
+
+			resp, err := client.DoProbes(context.Background(), tc.req)
+
+			if err == nil {
```

Comment:
We shouldn't be _conditionally_ making assertions based on whether our real implementation returned an error - we should be _asserting_ whether we _expect_ the real implementation to return an error.

Reply:
I had created a field in test cases for errors, no idea why I didn't use them. Thanks!